server/src/main/java/org/elasticsearch/cluster/routing/allocation/RoutingAllocation.java

```diff
@@ -42,15 +43,10 @@
 
     private final AllocationDeciders deciders;
 
+    @Nullable
     private final RoutingNodes routingNodes;
```

Comment:
We don't need to generate this when constructing the object. In some cases the routing nodes are never needed for the concrete cluster state passed in here because none of the allocators care about it. So we can selectively make it lazy like this and pass in the mutable variant as an override.